### DISCUSSION TITLE
[Q&A] Is DashSegmentedReader and Writer using In Memory ?

### DISCUSSION THREAD

**Discussion by DevataDev** [streamlink/streamlink#5361] [Q&A] [answered]
### Description

Hi All,
I am use Streamlink API for try to load DASH Stream, and i am using the DashStream object , but i notice is the memory usage for the FFMPEG ( FFmpegMux) process is increased from time to time, and when it's already running almost 1 - 2 hours the usage was almost 500+ MB.

Is it intended behaviour ?

Thank you

![image](https://github.com/streamlink/streamlink/assets/109707553/4f19542a-b2a7-4027-9999-85bde812e4ad)

**Comment by bastimeyer** [comment, marked as answer]
Streamlink has a separate ringbuffer (default is 16MiB) for each stream when muxing multiple streams via `FFMPEGMuxer`. The `FFMPEGMuxer` writes the data of each stream into a named pipe where the `ffmpeg` process is reading from. Depending on the bitrate of each stream, how ffmpeg is able to read the individual input streams (data arrival time of segmented streams), how it handles its input and output buffers while muxing, and how ffmpeg's output buffer eventually gets read by streamlink_cli or an implementer of Streamlink's Python API will make ffmpeg allocate more or less memory. Streamlink's `FFMPEGMuxer` doesn't set any ffmpeg input buffer sizes (which would lead to Streamlink's buffers being filled instead) or output buffer sizes. You can see the used ffmpeg arguments in Streamlink's debug log and there's also the option to print its verbose output to stderr or to a path on the filesystem.

- https://streamlink.github.io/cli.html#cmdoption-ringbuffer-size
- https://streamlink.github.io/cli.html#cmdoption-ffmpeg-verbose
- https://github.com/streamlink/streamlink/blob/5.5.1/src/streamlink/stream/ffmpegmux.py#L82

How FFmpeg handles its memory is of out of scope of Streamlink. Make sure that you're reading its output as soon as possible ([`FFMPEGMuxer.read()`](https://github.com/streamlink/streamlink/blob/5.5.1/src/streamlink/stream/ffmpegmux.py#L222-L223)) and that none of the input streams stall which would make ffmpeg wait on some data before it's able to mux the streams.

Otherwise, have a look at the code.

**Reply by DevataDev** [reply]
Thanks @bastimeyer will take a look on it 👍
